error_writer.go

```diff
@@ -41,7 +41,7 @@ type ErrorWriter struct {
 // RPC Content-Types in net/http middleware, you must pass the same
 // HandlerOptions to NewErrorWriter and any wrapped Connect handlers.
 func NewErrorWriter(opts ...HandlerOption) *ErrorWriter {
-	config := newHandlerConfig("", opts)
+	config := newHandlerConfig("", ^StreamType(0), opts)
```

Comment:
This seems like a problematic corner of the design. We're using the options here to know the codecs registered, so that we can understand the content-types that the handler supports. If we're now allowing codecs to be registered conditionally depending on the stream type, we'll have to make sure that the stream type we supply here is correct. I'm not sure how to do that.

The best options I can come up with are:

1. Add a `WithProcedureType` handler option and document that it only affects `ErrorWriter`.
2. Add an `ErrorWriter.SetProcedureType` method, which alters the default and reconfigures the error writer.

Neither of those feel great to me. Can anyone think of a nicer, backward-compatible alternative?

Reply:
`StreamType` is a bit mask with the zero type equal to `StreamTypeUnary`. Maybe we could have document an unknown sentinel type: `type StreamTypeUnknown StreamType = ^StreamType(0)`? I wanted the zero value not to match real ones to avoid confusing cases where applying options with a `spec.StreamType == StreamTypeUnary` could be applied to a `StreamTypeBidi` under error conditions. Probably not a issue though.

A non backward-compatible alternative is to move http handler middlerware onto the handler with an option like:
```go
type HandlerFunc func(w http.ResponseWriter, r *http.Request) error

func WithHandlerMiddleware(middleware func(HandlerFunc) HandlerFunc)
```
So each error handler is aware of the protocol and can deduce the `StreamType`. This maybe has too many sharp edges, but it would integrate nicely on the level of `Handler`.

Reply:
Would header middleware be fine for most applications?

```go
type HeaderFunc func(responseHeaders http.Header, requestHeader http.Header) error
```

Then the auth example [here](https://github.com/bufbuild/connect-go/blob/main/error_writer_example_test.go) could be written as:
```go
// NewAuthenticatedHandler middleware to authenticate a RPC request.
func NewAuthenticatedHandler(handler connect.HeaderFunc) connect.HeaderFunc {
	return func(responseHeader http.Header, requestHeader http.Header) error {
		// Dummy authentication logic.
		if requestHeader.Get("Token") == "super-secret" {
			return handler(responseHeader,  requestHeader)
		}
		// Send a protocol-appropriate error to RPC clients, so that they receive
		// the right code, message, and any metadata or error details.
		return connect.NewError(connect.CodeUnauthenticated, errors.New("invalid token"))
	})
}
```

Reply:
> unclear if it was meant to be used as a bitmask
I see what you mean, it looks like it would be checked for a streaming client or streaming response which then wouldn't make sense to have either bits set.
> crawl the set of associated options (no need to apply the predicate)
I'm don't understand how this works. You would have all the non optional options and union them with the conditional?

Reply:
> You would have all the non conditional options and union them with the conditional?

Correct. The error writer really just needs to know what content types are supported. Thinking about this more, I guess this would be problematic if the conditional options were used to _override_ a codec, such that the union of all codecs was ambiguous regarding which codec to use for a particular content type.

I think the only way we could handle that sort of setup would be to provide the `Spec` somehow to the error writer, similar to what Akshay suggested above. Though I'm now a little confused why the suggestions above focus solely on stream type, and not the whole spec, since the conditional options might have logic based on procedure name, too.

Reply:
I think we can easily get the procedure name from the path, given a valid request, but stream type would be harder. I don't think that information is available.

Reply:
> easily get the procedure name from the path

That's only true when [using protos](https://connect.build/docs/protocol#protobuf) (and gRPC conventions). But that's not actually a requirement in the Connect protocol spec. So, used with some other service description convention, the `Procedure-Name` production in the spec could map to something with fewer or more than two path components, in which case this assumption means we'd compute the wrong string.